src/type.cpp

```diff
@@ -158,6 +158,34 @@ std::string StructType::id() const noexcept {
   return id_;
 }
 
+std::size_t StructType::offset(const std::string& id) const {
+  std::size_t offset = 0;
+  for (auto i = std::size_t{0}, e = fields_.size(); i < e; ++i) {
+    const auto& field = fields_.at(i);
+    if (field->id == id) {
+      return offset;
+    }
+
+    offset += field->type->size();
+  }
+
+  throw "member not found in struct!";
+}
+
+std::size_t StructType::offset(const std::size_t index) const {
+  std::size_t offset = 0;
+  for (auto i = std::size_t{0}, e = fields_.size(); i < e; ++i) {
+    const auto& field = fields_.at(i);
+    if (i == index) {
+      return offset;
+    }
+
+    offset += field->type->size();
+  }
+
+  throw "member not found in struct!";
+}
```

Comment:
How about using `std::accumulate` to eliminate the loop and making the member check a special case? This means checking it with an additional `if` statement instead of making it the last statement. Consider the following pseudo-code:

```cpp
std::size_t StructType::offset(const std::size_t index) const {
  if (index >= fields_.size()) {
    throw ...
  }

  return std::accumulate(fields_.cbegin(), fields_.cend(),
    std::size_t{0}, [](auto&& size, auto&& field) {
      return size + field->type->size();
    });
}
```

This approach simplifies the code, clarifies the offset is the sum of sizes, and clearly handles the case where the index is out of bounds.